mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,279 @@
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+#define GEN_PASS_DEF_SPLITMULTIPLETAPESPASS
+#include "Catalyst/Transforms/Passes.h.inc"
+
+struct SplitMultipleTapesPass : public impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass> {
+    using impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass>::SplitMultipleTapesPassBase;
+
+    unsigned int countTapes(func::FuncOp func)
+    {
+        // Count the number of quantum.device operations in a function
+        unsigned int count = 0;
+        func->walk([&](Operation *op) {
+            if (op->getName().getStringRef() == "quantum.device") {
+                count++;
+            }
+        });
+        return count;
+    }
+
+    void CollectOperationsForEachTape(
+        func::FuncOp func, SmallVector<std::shared_ptr<SmallVector<Operation *>>> &OpsEachTape)
+    {
+        // During tracing, each tape starts with a qdevice_p primitive
+        // This means each tape starts with a quantum.device
+        // and ends with a quantum.device_release
+
+        // The structure of the FuncOp looks like the following:
+        // func.func @circuit(...) -> ... {...} {
+        // preprocessing
+        // quantum.device[...]
+        // tape 1
+        // quantum.device_release
+        // quantum.device[...]
+        // tape 2
+        // quantum.device_release
+        // ... more tapes
+        // post processing and return
+        // }
+
+        // This function parses the operations in the funcop and puts them into
+        // the container as
+        // OpsEachTape = [[preprocessing ops], [tape1 ops], [tape2 ops], ..., [postprocessing ops]]
+
+        for (size_t i = 0; i < OpsEachTape.size(); i++) {
+            OpsEachTape[i] = std::make_shared<SmallVector<Operation *>>();
+        }
+
+        // special: [0] for pre and [-1] for post processing
+        unsigned int cur_tape = 0;
+
+        func->walk([&](Operation *op) {
+            if (op == func) {
+                return; // don't visit the funcop itself
+            }
+            if (op->getName().getStringRef() == "quantum.device") {
+                cur_tape++;
+            }
+            OpsEachTape[cur_tape]->push_back(op);
+            if ((op->getName().getStringRef() == "quantum.device_release") &&
+                cur_tape == OpsEachTape.size() - 2) {
+                // reached post processing
+                cur_tape++;
+            }
+        });
+    }
+
+    LogicalResult CreateTapeFunction(std::shared_ptr<SmallVector<Operation *>> TapeOps,
+                                     SmallVector<Value> &NecessaryValuesFromEarlierTapes,
+                                     IRRewriter &builder, unsigned int tape_number,
+                                     func::FuncOp OriginalMultitapeFunc,
+                                     SmallVector<FailureOr<func::FuncOp>> &OutlinedFuncs)
+    {
+        assert(TapeOps); // nullptr protection
+
+        // 1. Identify the necessary return values
```

Comment:
Can you outline this into its own function?

Reply:
done